src/Course/StateT.hs

```diff
@@ -267,32 +268,38 @@ instance Functor f => Functor (OptionalT f) where
 instance Monad f => Applicative (OptionalT f) where
   pure = OptionalT . pure . Full
 
-  -- (<*>) :: OptionalT (f (Optional (a -> b))) -> OptionalT (f (Optional a)) -> OptionalT (f (Optional b))
-  -- ffunc :: f (Optional (a -> b))
-  -- fa    :: f (Optional a)
-  -- func <*> :: f _ -> f b
-  OptionalT ffunc <*> OptionalT fa = OptionalT $ (<*>) ffunc <*> _
-
-  -- OptionalT ffunc <*> OptionalT fa = OptionalT $ do
-  --   maybeFunc <- ffunc
-  --   maybeA <- fa
-  --   case maybeFunc of
-  --     Empty -> Empty
-  --     Full func -> case maybeA of
-  --       Empty -> Empty
-  --       Full a -> Full $ func a
-
-
-  -- OptionalT ffunc <*> OptionalT fa = onFull t -> f (Optional a) Optional t
-  --   Applicative f => (t -> f (Optional a)) -> Optional t -> f (Optional a)
-  -- theirs: OptionalT $ (<*>) <$> ffunc <*> fa (but doesn't pass all tests?)
+  OptionalT ffunc <*> OptionalT fa = OptionalT $ do
+    maybeFunc <- ffunc
+    case maybeFunc of
+      Empty     -> pure Empty
+      Full func -> do
+        a <- fa
+        pure $ func <$> a
+    -- REVIEW:
+    -- Above, you suggested solutions like these over more point-free ones like theirs
+    -- (OptionalT f <*> OptionalT a = OptionalT (lift2 (<*>) f a)). Does this fall into the same
```

Comment:
I think in this case their solution is cleaner: it straightforwardly states that `<*>` on `OptionalT` is `<*>` on `Optional` lifted through the underlying functor.
Note, however, that their solution is _not_ point-free: it has two points, `f` and `a`.

Reply:
I think I don't have a good intuitive sense of what `lift2` means, and my understanding of what `lift` means might be muddling it further. My understanding is that `lift` takes a value that is under one monad in a transformer stack and makes it compatible with the entire stack. Does `lift2` mean something similar, or is it completely different?